--- a/Lib/ctypes/util.py
+++ b/Lib/ctypes/util.py
@@ -450,7 +450,8 @@
 # https://man.openbsd.org/dl_iterate_phdr
 # https://docs.oracle.com/cd/E88353_01/html/E37843/dl-iterate-phdr-3c.html
 if (os.name == "posix" and
-    sys.platform not in {"darwin", "ios", "tvos", "watchos"}):
+    sys.platform not in {"darwin", "ios", "tvos", "watchos"} and
+    getattr(sys.implementation, "_multiarch", "") != "x86_64-filc-linux-musl"):
     import ctypes
     if hasattr((_libc := ctypes.CDLL(None)), "dl_iterate_phdr"):
 
